File: sdk-ts/src/sub/keys.ts
import {
  Account,
  Keypair,
  Networks,
  Operation,
  TransactionBuilder,
} from '@stellar/stellar-sdk';
import type { Credentials, Transport } from '../transport.js';

export interface IssuedChallenge {
  challengeHex: string;
  expiresAt: number;
}

export interface IssuedApiKey {
  keyId: string;
  secret: string;
  owner: string;
  tier: 'standard' | 'market_maker';
  createdAt: number;
}

export interface ApiKeyRecord {
  keyId: string;
  owner: string;
  tier: 'standard' | 'market_maker';
  label: string | null;
  createdAt: number;
  lastUsedAt: number | null;
  revokedAt: number | null;
}

export interface BetaStatus {
  /** True when key issuance is gated to a closed-beta allowlist. */
  gated: boolean;
  /** True when issuance is open, or the supplied address is on the allowlist. */
  allowed: boolean;
}

/**
 * Function the caller provides to sign 32 bytes (a transaction hash) with
 * the Stellar private key. The SDK never sees the secret. Examples:
 *
 *   import { Keypair } from '@stellar/stellar-sdk';
 *   const kp = Keypair.fromSecret('S...');
 *   const sign: ChallengeSigner = (data) => kp.sign(data);
 */
export type ChallengeSigner = (data: Buffer) => Buffer | Promise<Buffer>;

export interface KeysCreateOptions {
  address: string;
  signer: ChallengeSigner;
  label?: string;
  /**
   * Network passphrase used to construct the wrapping manageData tx.
   * Defaults to the public testnet (`Networks.TESTNET`). The value must
   * match the API gateway's network — pass `Networks.PUBLIC` for
   * mainnet deployments.
   */
  networkPassphrase?: string;
}

export class KeysApi {
  constructor(private readonly transport: Transport, private readonly credentials: Credentials | null) {}

  /** Public — whether key issuance is closed-beta gated (and if `address` is allowed). */
  async betaStatus(address?: string): Promise<BetaStatus> {
    return this.transport.request<BetaStatus>({
      path: '/v1/keys/beta-status',
      query: { address },
    });
  }

  async requestChallenge(address: string): Promise<IssuedChallenge> {
    return this.transport.request<IssuedChallenge>({
      method: 'POST',
      path: '/v1/keys/challenge',
      body: { address },
    });
  }

  /**
   * Exchange a SEP-10-style signed XDR for an API key. The `signature`
   * field on the server endpoint carries the full base64-encoded signed
   * transaction (legacy field name; kept for wire compatibility).
   */
  async exchange(input: {
    address: string;
    challenge: string;
    signedXdr: string;
    label?: string;
  }): Promise<IssuedApiKey> {
    return this.transport.request<IssuedApiKey>({
      method: 'POST',
      path: '/v1/keys',
      body: {
        address: input.address,
        challenge: input.challenge,
        signature: input.signedXdr,
        label: input.label,
      },
    });
  }

  /**
   * Convenience: request a challenge, wrap it as a manageData op on a
   * placeholder transaction, ask the caller's signer to sign the tx
   * hash, assemble a signed XDR, exchange it for an API key.
   *
   * This matches the gateway's verification path:
   *   1. Server parses the XDR.
   *   2. Asserts op[0] is manageData carrying the issued challenge.
   *   3. Verifies any tx signature against `tx.hash()` with the
   *      address's pubkey.
   */
  async create(input: KeysCreateOptions): Promise<IssuedApiKey> {
    const networkPassphrase = input.networkPassphrase ?? Networks.TESTNET;
    const challenge = await this.requestChallenge(input.address);

    // Build a sequence-0 placeholder tx whose only op is a manageData
    // carrying the challenge bytes. We use a random source account so
    // we don't need the user's actual account state.
    const placeholderSource = Keypair.random().publicKey();
    const account = new Account(placeholderSource, '0');
    const tx = new TransactionBuilder(account, {
      fee: '0',
      networkPassphrase,
    })
      .addOperation(
        Operation.manageData({
          name: 'noether-api auth',
          value: Buffer.from(challenge.challengeHex, 'hex'),
          source: input.address,
        }),
      )
      .setTimeout(0)
      .build();

    // Get the caller to sign the tx hash, then attach as a decorated
    // signature using the address's keypair hint.
    const txHash = tx.hash();
    const rawSig = await input.signer(txHash);
    const kp = Keypair.fromPublicKey(input.address);
    tx.addSignature(input.address, Buffer.from(rawSig).toString('base64'));
    void kp; // keep the import warning-free; hint is derived inside addSignature

    return this.exchange({
      address: input.address,
      challenge: challenge.challengeHex,
      signedXdr: tx.toXDR(),
      label: input.label,
    });
  }

  async list(): Promise<ApiKeyRecord[]> {
    this.requireAuth();
    const res = await this.transport.request<{ keys: ApiKeyRecord[] }>({
      path: '/v1/keys',
      credentials: this.credentials,
    });
    return res.keys;
  }

  async revoke(keyId: string): Promise<void> {
    this.requireAuth();
    await this.transport.request({
      method: 'DELETE',
      path: `/v1/keys/${encodeURIComponent(keyId)}`,
      credentials: this.credentials,
    });
  }

  private requireAuth(): void {
    if (!this.credentials) throw new Error('keys.list / keys.revoke require an authenticated client (pass credentials in NoetherClient options)');
  }
}

File: sdk-py/noether_sdk/sub/keys.py
from typing import TYPE_CHECKING, Awaitable, Callable

from ..errors import AuthError
from ..models import ApiKeyRecord, BetaStatus, IssuedApiKey, IssuedChallenge
from ..transport import Credentials, Transport

if TYPE_CHECKING:  # pragma: no cover - typing only
    from stellar_sdk import TransactionEnvelope


ChallengeSigner = Callable[[bytes], bytes | Awaitable[bytes]]

CHALLENGE_DATA_NAME = "noether-api auth"
TESTNET_PASSPHRASE = "Test SDF Network ; September 2015"


def _require_stellar_sdk():
    try:
        import stellar_sdk
    except ImportError as exc:  # pragma: no cover - exercised via the stellar extra
        raise RuntimeError(
            "keys.create requires the stellar-sdk package — install via "
            "`pip install 'noether-sdk[stellar]'`."
        ) from exc
    return stellar_sdk


def build_challenge_tx(
    address: str,
    challenge_hex: str,
    network_passphrase: str = TESTNET_PASSPHRASE,
) -> "TransactionEnvelope":
    """Wrap the gateway challenge as the value of a manageData op on a
    sequence-0 placeholder transaction (random source account, zero fee).

    This is the exact SEP-10-style envelope the gateway verifies:
    op[0] must be a manageData carrying the challenge bytes, and one of
    the attached signatures must verify against ``tx.hash()``.
    """
    stellar = _require_stellar_sdk()
    builder = stellar.TransactionBuilder(
        source_account=stellar.Account(stellar.Keypair.random().public_key, 0),
        network_passphrase=network_passphrase,
        base_fee=0,
    )
    builder.append_manage_data_op(
        data_name=CHALLENGE_DATA_NAME,
        data_value=bytes.fromhex(challenge_hex),
        source=address,
    )
    builder.add_time_bounds(0, 0)
    return builder.build()


class KeysApi:
    def __init__(self, transport: Transport, credentials: Credentials | None) -> None:
        self._transport = transport
        self._credentials = credentials

    async def beta_status(self, address: str | None = None) -> BetaStatus:
        """Public — whether key issuance is closed-beta gated (and if `address` is allowed)."""
        body = await self._transport.request(
            "GET", "/v1/keys/beta-status", params={"address": address}
        )
        return BetaStatus.model_validate(body)

    async def request_challenge(self, address: str) -> IssuedChallenge:
        body = await self._transport.request(
            "POST", "/v1/keys/challenge", json={"address": address}
        )
        return IssuedChallenge.model_validate(body)

    async def exchange(
        self,
        *,
        address: str,
        challenge: str,
        signed_xdr: str,
        label: str | None = None,
    ) -> IssuedApiKey:
        """Exchange a SEP-10-style signed XDR for an API key. The server's
        `signature` field carries the full base64-encoded signed
        transaction (legacy field name; kept for wire compatibility).
        """
        body = await self._transport.request(
            "POST",
            "/v1/keys",
            json={
                "address": address,
                "challenge": challenge,
                "signature": signed_xdr,
                "label": label,
            },
        )
        return IssuedApiKey.model_validate(body)

    async def create(
        self,
        *,
        address: str,
        signer: ChallengeSigner,
        label: str | None = None,
        network_passphrase: str = TESTNET_PASSPHRASE,
    ) -> IssuedApiKey:
        """Convenience: request a challenge, wrap it as a manageData op on
        a placeholder transaction, ask the caller's signer to sign the tx
        hash, assemble a signed XDR, exchange it for an API key.

        This matches the gateway's verification path:
          1. Server parses the XDR.
          2. Asserts op[0] is manageData carrying the issued challenge.
          3. Verifies any tx signature against ``tx.hash()`` with the
             address's pubkey.

        The passphrase must match the gateway's network — pass the
        public-network passphrase for mainnet deployments.
        """
        stellar = _require_stellar_sdk()
        challenge = await self.request_challenge(address)
        tx = build_challenge_tx(address, challenge.challenge_hex, network_passphrase)
        sig = signer(tx.hash())
        if hasattr(sig, "__await__"):
            sig = await sig  # type: ignore[assignment]
        hint = stellar.Keypair.from_public_key(address).signature_hint()
        tx.signatures.append(stellar.DecoratedSignature(hint, bytes(sig)))
        return await self.exchange(
            address=address,
            challenge=challenge.challenge_hex,
            signed_xdr=tx.to_xdr(),
            label=label,
        )

    async def list(self) -> list[ApiKeyRecord]:
        self._require_auth()
        body = await self._transport.request("GET", "/v1/keys", credentials=self._credentials)
        return [ApiKeyRecord.model_validate(k) for k in body.get("keys", [])]

    async def revoke(self, key_id: str) -> None:
        self._require_auth()
        await self._transport.request(
            "DELETE", f"/v1/keys/{key_id}", credentials=self._credentials
        )

    def _require_auth(self) -> None:
        if self._credentials is None:
            raise AuthError(
                "keys.list / keys.revoke require an authenticated client",
                status=401,
                body=None,
                url="/v1/keys",
            )
